test/envtest/watch.go

```diff
@@ -0,0 +1,48 @@
+package envtest
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// watchFor watches for an event of type eventType using the provided watch.Interface.
+// It returns when either context is done - and then it marks the test as failed -
+// or when the event has been received and predicate returned true.
+func watchFor[
+	T metav1.Object,
+](
+	t *testing.T,
+	ctx context.Context,
+	w watch.Interface,
```

Comment:
Do you think we could make this helper also run the watch itself instead of relying on the caller to pass it down? This way we could take care of cleaning the watch up in this function, making the caller code leaner.

Reply:
Yes, but the problem is that we first need to set up the watch and then we can perform an action that the effect of which we'll be catching with the watch.

When we encapsulate the watch within the `watchFor` we can't do it. I've tried adding a `preactions ... func(*testing.T, context.Context)` argument to allow optional actions which would be called between the watch set up and the for loop but I'm not convinced about this. 

LMYWKT.

Reply:
Ah, I see. It makes sense to me now. `preactions` doesn't sound intuitive to me either. 

Possibly, we could have a separate wrapper that creates a watch and sets up its cleanup to reduce boilerplate as much as we can. So instead of 
```go
	w, err := clientWithWatch.Watch(ctx, &configurationv1alpha1.KongPluginBindingList{}, client.InNamespace(ns.Name))
	require.NoError(t, err)
	t.Cleanup(func() { w.Stop() })
```
we'd have
```go
	w := setupWatch(ctx, t, clientWithWatch, &configurationv1alpha1.KongPluginBindingList{}, client.InNamespace(ns.Name))
```

Reply:
Makes sense 👍 c375eadd65916df85a3ffeb35ff775b9103f8997

Let me know if that looks reasonable.